### PR TITLE
added poc for progressive update mechanism

### PR DESCRIPTION
**Pull request by IAmMorrow**
this PR feature a progressive update mechanism and a remote config system.

On app boot, the following data structure is fetched from github and loaded in memory: 

> https://github.com/LedgerHQ/ledger-live-desktop/blob/config/config.json

it then calculate a deterministic random local update threshold using the userId. If the threshold is inferior than the global one provided in the remote config, the update banner will appear. It is possible to define one global update threshold per platform (darwin, win32, linux ...).

Notes:

- the whole system is optional and if the config doesn't have the required data, the update will just appear normaly
- the deterministic random is based on a ${userId}-${version} seed, this way it's not always the same users that get the update first !

How to test ?

To test, you have to modify your package.json version number to something anterior to the last release (2.0.0 for instance), then you have to package the app (yarn dist).
Run the packaged app, and an update banner will appear (or not, 1 chance out of two with the actual config.json file)

Reset your userId in the localstorage if you want to reset the odds ;)

### CODE REVIEW

File: src/helpers/user.js [hunk 1]
```diff
@@ -29,3 +31,16 @@ export const getUserId = () => {
   }
   throw new Error("user is only to be called from renderer");
 };
+
+export const shouldUpdateYet = (version: string, remoteConfig: any) => {
+  const userId = getUserId();
+  const generateNumber = seedrandom(`${userId}-${version}`);
```
Comment (gre) [inline src/helpers/user.js]: `Prando` is the lib we use on live-common that is less problematic, especially on diff JS env (seedrandom does not run on RN for instance. not that we would use here but i kinda feel this code could belong to https://github.com/LedgerHQ/ledger-live-common/blob/master/src/user.js which centralize the "derives from user id" logic)

File: src/renderer/components/Updater/Banner.js [hunk 2]
```diff
@@ -51,12 +53,18 @@ const CONTENT_BY_STATUS = (quitAndInstall, reDownload, progress): { [string]: Co
...
 const UpdaterTopBanner = () => {
   const context = useContext(UpdaterContext);
+  const remoteConfig = useRemoteConfig();
 
   const reDownload = useCallback(() => {
     openURL(urls.liveHome);
   }, []);
 
-  if (context) {
+  if (context && remoteConfig && context.version) {
```
Comment (gre) [inline src/renderer/components/Updater/Banner.js]: ⚠️  this makes the release not resilient to network down problem or problem in the fetching config.

I think by default we always should suggest the auto update when fetching config fails. the challenge might be to not having it "flash" during the loading phase, so you might want to track if it's loading/error/loaded.

Comment (IAmMorrow) [reply src/renderer/components/Updater/Banner.js]: I fixed it by adding a "lastUpdatedAt" param that ensure that we tried at least once to fetch the remote config, sucessfully or not.

File: src/helpers/user.js (hunk 1 above)
Comment (IAmMorrow) [reply src/helpers/user.js]: I replaced seedrandom with Prando

**Comment by gre** [comment]
:pikasurprised:

**Comment by gre** [comment]
that's a good feedback that i didn't anticipate, i think it's because I might have intentionally said "do it on ui side" as i thought was easiest and i didn't consider the concerned raised here.

**Comment by IAmMorrow** [comment]
It's okay I will implement an "on user action" download that will settle this

**Comment by IAmMorrow** [comment]
I added a banner that prompt "Ledger live {version} available for update" with a "Download now" button

File: package.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
   "name": "ledger-live-desktop",
   "productName": "Ledger Live",
   "description": "Ledger Live - Desktop",
-  "version": "2.1.0",
+  "version": "2.0.0",
```
Comment (MortalKastor) [inline package.json]:
```suggestion
  "version": "2.1.0",
```

(unless it was on purpose to help QA?)

Comment (IAmMorrow) [reply package.json]: not on purpose, I forgot it, I'm getting really slow

File: src/main/updater/index.js [hunk 4]
```diff
@@ -18,6 +18,14 @@ export default (type: string) => {
       }
       break;
 
+    case "download-update":
+      if (!updater.downloadUpdate) {
+        logger.error(`Auto-update error: quitAndInstall called before init`);
```
[the remark this answers is not in the document]
Comment (IAmMorrow) [reply src/main/updater/index.js]: damn me

**Pull request merged**